fncas/linux_native_jit/benchmark.cc

```diff
@@ -0,0 +1,200 @@
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+// A simple real-life-ish optimization problem to measure the performance of the native Linux JIT.
+//
+// The problem is effectively a sparse linear programming one with a softmax-based cost function.
+// For `--n` variables and `--m` training examples, with sparseness coefficient `--k`, so that
+// exactly `m * k` variables are set per each training example.
+//
+// Each variable has its own `true` value, which is a random integer between one and nine.
+// Each training example has its own `true` right-hand-side value, which is a sum of values per variables set
+// for this training example.
+//
+// Now, for the cost function, instead of a min-squares loss (which would make the above an LP problem, for which
+// the conjugate gradient descent method is a perfect fit), this example code uses:
+//
+// total_penalty = sum(penalty_per_point[p])
+// penalty_per_point[p] = softmax(true_value[p] - computed_value[p], 0) + softmax(computed_value[p] - true_value[p], 0)
+// where softmax(a, b) = log(exp(a) + exp(b)), and thus softmax(x, 0) = log(exp(x) + 1)
+//
+// Since no random noise is introduced for the picture, the lowest possible `penalty_per_point[p]` is `2 * log(2)`.
+// To keep things simple, in this example this value is subtracted from the cost function,
+// so that ultimately it optimizes towards zero.
+
+#include "../fncas.h"
+
+#include "../../bricks/dflags/dflags.h"
+#include "../../bricks/util/random.h"
+
+DEFINE_int64(n, 250, "The number of training examples.");
+DEFINE_int64(m, 100, "The number of variables to train on.");
+DEFINE_double(k, 0.1, "The fraction of variables used per training example.");
+
+DEFINE_string(function, "softmaxes", "The cost function to optimize for, `softmaxes/l1/l2`.");
+DEFINE_string(optimizer, "jit", "The gradient evaluation technique to use `jit|as|clang|slow`.");
+DEFINE_uint32(max_iterations, 10000, "The maximum number of iterations to make.");
+
+DEFINE_bool(dump, false, "Set to dump the input data and the optimization result.");
+DEFINE_bool(log, false, "Set to see the log of optimization iterations.");
+
+DEFINE_int32(random_seed, 42, "The random seed to use. System- and compiler-dependent still.");
+DEFINE_int64(a, 1, "The left end of the range to uniformly draw the true variable values from.");
+DEFINE_int64(b, 9, "The [included] right end of the range to uniformly draw the true variable values from.");
+
+struct Data {
+  size_t const n;
+  size_t const m;
+  double const k;
+  size_t const q;                                  // Just `m * k` rounded to an integer.
+  std::vector<std::vector<size_t>> sparse_matrix;  // Sorted per training example.
```

Comment:
Let's stick to our convention of using explicitly-sized data types everywhere where we deal with serialization. Actually, it doesn't compile on my clang unless `size_t` is replaced with `uint64_t`.

Reply:
:+1: debc4015